projects/stretch_continual/stretch_continual/envs/stretch_demo_base_env.py

```diff
@@ -0,0 +1,312 @@
+        negative_roll = positive_roll - (2 * np.pi)
+        current_roll = current_joints[HelloStretchIdx.WRIST_ROLL]
+
+        if np.abs(current_roll - positive_roll) < np.abs(current_roll - negative_roll):
+            ros_pose[HelloStretchIdx.WRIST_ROLL] = positive_roll
+        else:
+            ros_pose[HelloStretchIdx.WRIST_ROLL] = negative_roll
+
+        return ros_pose
+
+    def construct_observation(
+        self,
+        image,
+        depth,
+        pose,
+        color_camera_info,
+        depth_camera_info,  # Intentionally unused, to make it clearer how to pass this information if desired
+        camera_pose,
+        camera_info_in_state,
+        model,
+        context_observation=None,
+    ):
+        """
+        Expects {color, depth}_camera_info to be a dict of D (5,), K (3x3), R (3x3), P (3x4)
+        Expects camera_pose to be shape (1, 4, 4)
+        """
+        # The ee pose from fk doesn't include the offset to the gripper properly -- include it
+        ee_pos, ee_rot = self.gripper_fk(model, pose)
+
+        gripper_pose = pose[HelloStretchIdx.GRIPPER]
+        adjusted_pose = np.concatenate(
+            (ee_pos, ee_rot, np.array([gripper_pose, 0])), axis=-1  # TODO: remove the 0
+        )
+
+        # Just passing color camera info along for now, assuming depth is consistent
+        if camera_info_in_state:
+            color_camera_state = np.concatenate(
+                (
+                    color_camera_info["D"],
+                    color_camera_info["K"].flatten(),
+                    color_camera_info["R"].flatten(),
+                    color_camera_info["P"].flatten(),
+                )
+            )
+            camera_pose = camera_pose.flatten()
+            adjusted_pose = np.concatenate(
+                (adjusted_pose, color_camera_state, camera_pose)
+            )
+
+        depth = np.clip(depth, a_min=None, a_max=CAMERA_FAR_PLANE * CAMERA_SCALE_CONST)
+        depth_scale = 2**8 / (
+            CAMERA_SCALE_CONST * CAMERA_FAR_PLANE
+        )  # Rescale to get the most we can out of our bits
+        depth = depth * depth_scale
+
+        # Convert from uint16 to uint8 (2**8/2**16 = 1/2**8)
+        combined_image = np.concatenate(
+            (image, np.expand_dims(depth, axis=-1)), axis=-1
+        ).astype(np.uint8)
+
+        if context_observation is not None:
+            combined_image = np.concatenate(
+                (context_observation["image"], combined_image), axis=-1
+            )
+            adjusted_pose = np.concatenate(
+                (context_observation["state_vector"], adjusted_pose), axis=-1
+            )
+
+        return {"image": combined_image, "state_vector": adjusted_pose}
+
+    def get_stretch_obs_and_action_space(self, camera_info_in_state):
+        channels = 4
+        state_size = 9  # TODO: should be 8...keeping for now so I can keep using my trained model
+        state_size += 51 if camera_info_in_state else 0
```

Comment:
weird to have this in state isn't it?

Reply:
fair enough though

Reply:
Yeah, it is, but it is tremendously convenient. It's kind of nice to just pass through "image" and "state" and kind of pile everything into state I might need, because then IMPALA handles buffers/etc really cleanly, without having to put in more env-specific code. It's one of those things that I might shift as the use cases become more clear.

Reply:
yeah this makes sense